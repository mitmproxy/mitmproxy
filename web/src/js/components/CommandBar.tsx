import React, { useEffect, useRef, useState } from "react";
import classnames from "classnames";
import { fetchApi, runCommand } from "../utils";
import Filt from "../filt/command";

type CommandParameter = {
    name: string;
    type: string;
    kind: string;
};

type Command = {
    help?: string;
    parameters: CommandParameter[];
    return_type: string | undefined;
    signature_help: string;
};

type AllCommands = {
    [name: string]: Command;
};

type CommandHelpProps = {
    nextArgs: string[];
    currentArg: number;
    help: string;
    description: string;
    availableCommands: string[];
};

type CommandResult = {
    command: string;
    result: string;
};

type ResultProps = {
    results: CommandResult[];
};

function getAvailableCommands(
    commands: AllCommands,
    input: string = ""
): string[] {
    if (!commands) return [];
    let availableCommands: string[] = [];
    for (const [command, args] of Object.entries(commands)) {
        if (command.startsWith(input)) {
            availableCommands.push(command);
        }
    }
    return availableCommands;
}

export function Results({ results }: ResultProps) {
    const resultElement = useRef<HTMLDivElement>(null!);

    useEffect(() => {
        if (resultElement) {
            resultElement.current.addEventListener(
                "DOMNodeInserted",
                (event) => {
                    const target = event.currentTarget as Element;
                    target.scroll({
                        top: target.scrollHeight,
                        behavior: "auto",
                    });
                }
            );
        }
    }, []);

    return (
        <div className="command-result" ref={resultElement}>
            {results.map((result, i) => (
                <div key={i}>
                    <div>
                        <strong>$ {result.command}</strong>
                    </div>
                    {result.result}
                </div>
            ))}
        </div>
    );
}

export function CommandHelp({
    nextArgs,
    currentArg,
    help,
    description,
    availableCommands,
}: CommandHelpProps) {
    let argumentSuggestion: JSX.Element[] = [];
    for (let i: number = 0; i < nextArgs.length; i++) {
        if (i == currentArg) {
            argumentSuggestion.push(<mark key={i}>{nextArgs[i]}</mark>);
            continue;
        }
        argumentSuggestion.push(<span key={i}>{nextArgs[i]} </span>);
    }
    return (
        <div className="argument-suggestion popover top">
            <div className="arrow" />
            <div className="popover-content">
                {argumentSuggestion.length > 0 && (
                    <div>
                        <strong>Argument suggestion:</strong>{" "}
                        {argumentSuggestion}
                    </div>
                )}
                {help?.includes("->") && (
                    <div>
                        <strong>Signature help: </strong>
                        {help}
                    </div>
                )}
                {description && <div># {description}</div>}
                <div>
                    <strong>Available Commands: </strong>
                    <p className="available-commands">
                        {JSON.stringify(availableCommands)}
                    </p>
                </div>
            </div>
        </div>
    );
}

export default function CommandBar() {
    const [input, setInput] = useState<string>("");
    const [originalInput, setOriginalInput] = useState<string>("");
    const [currentCompletion, setCurrentCompletion] = useState<number>(0);
    const [completionCandidate, setCompletionCandidate] = useState<string[]>(
        []
    );

    const [availableCommands, setAvailableCommands] = useState<string[]>([]);
    const [allCommands, setAllCommands] = useState<AllCommands>({});
    const [nextArgs, setNextArgs] = useState<string[]>([]);
    const [currentArg, setCurrentArg] = useState<number>(0);
    const [signatureHelp, setSignatureHelp] = useState<string>("");
    const [description, setDescription] = useState<string>("");

    const [results, setResults] = useState<CommandResult[]>([]);
    const [history, setHistory] = useState<string[]>([]);
    const [currentPos, setCurrentPos] = useState<number | undefined>(undefined);

    useEffect(() => {
        fetchApi("/commands", { method: "GET" })
            .then((response) => response.json())
            .then((data: AllCommands) => {
                setAllCommands(data);
                setCompletionCandidate(getAvailableCommands(data));
                setAvailableCommands(Object.keys(data));
            })
            .catch((e) => console.error(e));
    }, []);

    useEffect(() => {
        runCommand("commands.history.get")
            .then((ret) => {
                setHistory(ret.value);
            })
            .catch((e) => console.error(e));
    }, []);

    const parseCommand = (originalInput: string, input: string) => {
        const parts: string[] = Filt.parse(input);
        const originalParts: string[] = Filt.parse(originalInput);

        setSignatureHelp(allCommands[parts[0]]?.signature_help);
        setDescription(allCommands[parts[0]]?.help || "");

        setCompletionCandidate(
            getAvailableCommands(allCommands, originalParts[0])
        );
        setAvailableCommands(getAvailableCommands(allCommands, parts[0]));

        const nextArgs: string[] = allCommands[parts[0]]?.parameters.map(
            (p) => p.name
        );

        if (nextArgs) {
            setNextArgs([parts[0], ...nextArgs]);
            setCurrentArg(parts.length - 1);
        }
    };

    const onChange = (e) => {
        setInput(e.target.value);
        setOriginalInput(e.target.value);
        setCurrentCompletion(0);
    };

    const onKeyDown = (e) => {
        if (e.key === "Enter") {
            const [cmd, ...args] = Filt.parse(input);

            setHistory([...history, input]);
            runCommand("commands.history.add", input).catch(() => 0);

            fetchApi
                .post(`/commands/${cmd}`, { arguments: args })
                .then((response) => response.json())
                .then((data) => {
                    setCurrentPos(undefined);
                    setNextArgs([]);
                    setResults([
                        ...results,
                        {
                            command: input,
                            result: JSON.stringify(data.value || data.error),
                        },
                    ]);
                })
                .catch((e) => {
                    setCurrentPos(undefined);
                    setNextArgs([]);
                    setResults([
                        ...results,
                        {
                            command: input,
                            result: e.toString(),
                        },
                    ]);
                });

            setSignatureHelp("");
            setDescription("");

            setInput("");
            setOriginalInput("");

            setCurrentCompletion(0);
            setCompletionCandidate(availableCommands);
        }
        if (e.key === "ArrowUp") {
            let nextPos;
            if (currentPos === undefined) {
                nextPos = history.length - 1;
            } else {
                nextPos = Math.max(0, currentPos - 1);
            }
            setInput(history[nextPos]);
            setOriginalInput(history[nextPos]);
            setCurrentPos(nextPos);
        }
        if (e.key === "ArrowDown") {
            if (currentPos === undefined) {
                return;
            } else if (currentPos == history.length - 1) {
                setInput("");
                setOriginalInput("");
                setCurrentPos(undefined);
            } else {
                const nextPos = currentPos + 1;
                setInput(history[nextPos]);
                setOriginalInput(history[nextPos]);
                setCurrentPos(nextPos);
            }
        }
        if (e.key === "Tab") {
            setInput(completionCandidate[currentCompletion]);
            setCurrentCompletion(
                (currentCompletion + 1) % completionCandidate.length
            );
            e.preventDefault();
        }
        e.stopPropagation();
    };

    const onKeyUp = (e) => {
        if (!input) {
            setAvailableCommands(Object.keys(allCommands));
            return;
        }
        parseCommand(originalInput, input);
        e.stopPropagation();
    };

    return (
        <div className="command">
            <div className="command-title">Command Result</div>
            <Results results={results} />
            <CommandHelp
                nextArgs={nextArgs}
                currentArg={currentArg}
                help={signatureHelp}
                description={description}
                availableCommands={availableCommands}
            />
            <div className={classnames("command-input input-group")}>
                <span className="input-group-addon">
                    <i className={"fa fa-fw fa-terminal"} />
                </span>
                <input
                    type="text"
                    placeholder="Enter command"
                    className="form-control"
                    value={input || ""}
                    onChange={onChange}
                    onKeyDown={onKeyDown}
                    onKeyUp={onKeyUp}
                />
            </div>
        </div>
    );
}
