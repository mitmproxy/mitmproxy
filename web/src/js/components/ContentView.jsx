import React, { Component, PropTypes } from 'react'
import { connect } from 'react-redux'
import * as ContentViews from './ContentView/ContentViews'
import * as MetaViews from './ContentView/MetaViews'
import ShowFullContentButton from './ContentView/ShowFullContentButton'


import { displayLarge, updateEdit } from '../ducks/ui/flow'

ContentView.propTypes = {
    // It may seem a bit weird at the first glance:
    // Every view takes the flow and the message as props, e.g.
    // <Auto flow={flow} message={flow.request}/>
    flow: React.PropTypes.object.isRequired,
    message: React.PropTypes.object.isRequired,
}

ContentView.isContentTooLarge = msg => msg.contentLength > 1024 * 1024 * (ContentViews.ViewImage.matches(msg) ? 10 : 0.2)

function ContentView(props) {
    const { flow, message, contentView, isDisplayLarge, displayLarge, onContentChange, readonly } = props

    if (message.contentLength === 0 && readonly) {
        return <MetaViews.ContentEmpty {...props}/>
    }

    if (message.contentLength === null && readonly) {
        return <MetaViews.ContentMissing {...props}/>
    }

    if (!isDisplayLarge && ContentView.isContentTooLarge(message)) {
        return <MetaViews.ContentTooLarge {...props} onClick={displayLarge}/>
    }

    const View = ContentViews[contentView] || ContentViews['ViewServer']
    return (
        <div className="contentview">
            <View flow={flow} message={message} contentView={contentView} readonly={readonly} onChange={onContentChange}/>
            <ShowFullContentButton/>
        </div>
    )
}

export default connect(
    state => ({
        contentView: state.ui.flow.contentView,
        isDisplayLarge: state.ui.flow.displayLarge,
    }),
    {
        displayLarge,
        updateEdit
    }
)(ContentView)
