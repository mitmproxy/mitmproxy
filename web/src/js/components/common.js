import React from "react"
import ReactDOM from "react-dom"
import _ from "lodash"

export var Router = {
    contextTypes: {
        location: React.PropTypes.object,
        router: React.PropTypes.object.isRequired
    },
    updateLocation: function (pathname, queryUpdate) {
        if (pathname === undefined) {
            pathname = this.context.location.pathname;
        }
        var query = this.context.location.query;
        if (queryUpdate !== undefined) {
            for (var i in queryUpdate) {
                if (queryUpdate.hasOwnProperty(i)) {
                    query[i] = queryUpdate[i] || undefined; //falsey values shall be removed.
                }
            }
        }
        this.context.router.replace({pathname, query});
    },
    getQuery: function () {
        // For whatever reason, react-router always returns the same object, which makes comparing
        // the current props with nextProps impossible. As a workaround, we just clone the query object.
        return _.clone(this.context.location.query);
    }
};

export var Splitter = React.createClass({
    getDefaultProps: function () {
        return {
            axis: "x"
        };
    },
    getInitialState: function () {
        return {
            applied: false,
            startX: false,
            startY: false
        };
    },
    onMouseDown: function (e) {
        this.setState({
            startX: e.pageX,
            startY: e.pageY
        });
        window.addEventListener("mousemove", this.onMouseMove);
        window.addEventListener("mouseup", this.onMouseUp);
        // Occasionally, only a dragEnd event is triggered, but no mouseUp.
        window.addEventListener("dragend", this.onDragEnd);
    },
    onDragEnd: function () {
        ReactDOM.findDOMNode(this).style.transform = "";
        window.removeEventListener("dragend", this.onDragEnd);
        window.removeEventListener("mouseup", this.onMouseUp);
        window.removeEventListener("mousemove", this.onMouseMove);
    },
    onMouseUp: function (e) {
        this.onDragEnd();

        var node = ReactDOM.findDOMNode(this);
        var prev = node.previousElementSibling;
        var next = node.nextElementSibling;

        var dX = e.pageX - this.state.startX;
        var dY = e.pageY - this.state.startY;
        var flexBasis;
        if (this.props.axis === "x") {
            flexBasis = prev.offsetWidth + dX;
        } else {
            flexBasis = prev.offsetHeight + dY;
        }

        prev.style.flex = "0 0 " + Math.max(0, flexBasis) + "px";
        next.style.flex = "1 1 auto";

        this.setState({
            applied: true
        });
        this.onResize();
    },
    onMouseMove: function (e) {
        var dX = 0, dY = 0;
        if (this.props.axis === "x") {
            dX = e.pageX - this.state.startX;
        } else {
            dY = e.pageY - this.state.startY;
        }
        ReactDOM.findDOMNode(this).style.transform = "translate(" + dX + "px," + dY + "px)";
    },
    onResize: function () {
        // Trigger a global resize event. This notifies components that employ virtual scrolling
        // that their viewport may have changed.
        window.setTimeout(function () {
            window.dispatchEvent(new CustomEvent("resize"));
        }, 1);
    },
    reset: function (willUnmount) {
        if (!this.state.applied) {
            return;
        }
        var node = ReactDOM.findDOMNode(this);
        var prev = node.previousElementSibling;
        var next = node.nextElementSibling;

        prev.style.flex = "";
        next.style.flex = "";

        if (!willUnmount) {
            this.setState({
                applied: false
            });
        }
        this.onResize();
    },
    componentWillUnmount: function () {
        this.reset(true);
    },
    render: function () {
        var className = "splitter";
        if (this.props.axis === "x") {
            className += " splitter-x";
        } else {
            className += " splitter-y";
        }
        return (
            <div className={className}>
                <div onMouseDown={this.onMouseDown} draggable="true"></div>
            </div>
        );
    }
});

export const ToggleComponent = (props) =>
    <div
      className={"btn " + (props.checked ? "btn-primary" : "btn-default")}
      onClick={props.onToggleChanged}>
      <span><i className={"fa " + (props.checked ? "fa-check-square-o" : "fa-square-o")}></i> {props.name}</span>
    </div>

ToggleComponent.propTypes = {
    name: React.PropTypes.string.isRequired,
    onToggleChanged: React.PropTypes.func.isRequired
}