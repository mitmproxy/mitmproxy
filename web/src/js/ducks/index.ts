import {
    AnyAction,
    applyMiddleware,
    combineReducers,
    compose,
    createStore as createReduxStore,
    PreloadedState,
    Store
} from "redux"
import eventLog from "./eventLog"
import flows from "./flows"
import ui from "./ui/index"
import connection from "./connection"
import options from './options'
import commandBar from "./commandBar";
import thunk, {ThunkAction, ThunkDispatch, ThunkMiddleware} from "redux-thunk";
import {TypedUseSelectorHook, useDispatch, useSelector} from "react-redux";
import backendState from "./backendState";
import options_meta from "./options_meta";


// @ts-ignore
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const rootReducer = combineReducers({
    commandBar,
    eventLog,
    flows,
    connection,
    ui,
    options,
    options_meta,
    backendState,
});
export type RootState = ReturnType<typeof rootReducer>


export type AppThunk<ReturnType = void> = ThunkAction<ReturnType,
    RootState,
    unknown,
    AnyAction>

export const createAppStore = (preloadedState?: PreloadedState<RootState>): Store<RootState> => {
    return createReduxStore(
        rootReducer,
        preloadedState,
        composeEnhancers(applyMiddleware(
            thunk as ThunkMiddleware<RootState>
        )))
};

export const store = createAppStore(undefined);

// this would be correct, but PyCharm bails on it
// export type AppDispatch = typeof store.dispatch
// instead:
export type AppDispatch = ThunkDispatch<RootState, void, AnyAction>
export const useAppDispatch = () => useDispatch<AppDispatch>()
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
