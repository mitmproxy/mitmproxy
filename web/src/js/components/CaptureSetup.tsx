import * as React from "react";
import {useEffect, useRef} from "react";
import {useAppSelector} from "../ducks";
import {ServerInfo} from "../ducks/backendState";
import {formatAddress} from "../utils";
import QRCode from 'qrcode';

export default function CaptureSetup() {
    const servers = useAppSelector(state => state.backendState.servers);

    let configure_action_text;
    if (servers.length === 0) {
        configure_action_text = "";
    } else if (servers.length === 1) {
        configure_action_text = "Configure your client to use the following proxy server:";
    } else {
        configure_action_text = "Configure your client to use one of the following proxy servers:";
    }

    return <div style={{padding: "1em 2em"}}>

        <h3>mitmproxy is running.</h3>
        <p>
            No flows have been recorded yet.<br/>
            {configure_action_text}
        </p>
        <ul className="fa-ul">
            {servers.map((server, i) => <li key={server.full_spec}><ServerDescription {...server}/></li>)}
        </ul>
        {/*
        <p>You can also start additional servers:</p>
        <ul>
            <li>TODO</li>
        </ul>
        */}

    </div>
}

export function ServerDescription(
    {
        description,
        listen_addrs,
        last_exception,
        is_running,
        full_spec,
        wireguard_conf,
    }: ServerInfo
) {
    const qrCode = useRef(null);
    useEffect(() => {
        if (wireguard_conf && qrCode.current)
            QRCode.toCanvas(qrCode.current, wireguard_conf, {margin: 0, scale: 3});
    }, [wireguard_conf]);

    let listen_str;
    const all_same_port = listen_addrs.length === 1 || (listen_addrs.length === 2 && listen_addrs[0][1] === listen_addrs[1][1]);
    const unbound = listen_addrs.every(addr => ["::", "0.0.0.0"].includes(addr[0]));
    if (all_same_port && unbound) {
        listen_str = formatAddress(["*", listen_addrs[0][1]]);
    } else {
        listen_str = listen_addrs.map(formatAddress).join(" and ");
    }
    description = description[0].toUpperCase() + description.substr(1);
    let desc, icon;
    if (last_exception) {
        icon = "fa-exclamation text-error"
        desc = <>{description} ({full_spec}):<br/>{last_exception}</>;
    } else if (!is_running) {
        icon = "fa-pause text-warning"
        desc = <>{description} ({full_spec})</>
    } else {
        icon = "fa-check text-success"
        desc = `${description} listening at ${listen_str}.`

        if (wireguard_conf) {
            desc = <>
                {desc}
                <div className="wireguard-config">
                    <pre>{wireguard_conf}</pre>
                    <canvas ref={qrCode}/>
                </div>
            </>;
        }

    }
    return <><i className={`fa fa-li ${icon}`}/>{desc}</>;
}
