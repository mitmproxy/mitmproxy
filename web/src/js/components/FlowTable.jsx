import React from 'react'
import PropTypes from 'prop-types'
import ReactDOM from 'react-dom'
import { connect } from 'react-redux'
import shallowEqual from 'shallowequal'
import AutoScroll from './helpers/AutoScroll'
import { calcVScroll } from './helpers/VirtualScroll'
import FlowTableHead from './FlowTable/FlowTableHead'
import FlowRow from './FlowTable/FlowRow'
import Filt from "../filt/filt"
import * as flowsActions from '../ducks/flows'


class FlowTable extends React.Component {

    static propTypes = {
        selectFlow: PropTypes.func.isRequired,
        flows: PropTypes.array.isRequired,
        rowHeight: PropTypes.number,
        highlight: PropTypes.string,
        selected: PropTypes.object,
    }

    static defaultProps = {
        rowHeight: 32,
    }

    constructor(props, context) {
        super(props, context)

        this.state = { vScroll: calcVScroll() }
        this.onViewportUpdate = this.onViewportUpdate.bind(this)
    }

    UNSAFE_componentWillMount() {
        window.addEventListener('resize', this.onViewportUpdate)
    }

    UNSAFE_componentWillUnmount() {
        window.removeEventListener('resize', this.onViewportUpdate)
    }

    componentDidUpdate() {
        this.onViewportUpdate()

        if (!this.shouldScrollIntoView) {
            return
        }

        this.shouldScrollIntoView = false

        const { rowHeight, flows, selected } = this.props
        const viewport = ReactDOM.findDOMNode(this)
        const head = ReactDOM.findDOMNode(this.refs.head)

        const headHeight = head ? head.offsetHeight : 0

        const rowTop = (flows.indexOf(selected) * rowHeight) + headHeight
        const rowBottom = rowTop + rowHeight

        const viewportTop = viewport.scrollTop
        const viewportHeight = viewport.offsetHeight

        // Account for pinned thead
        if (rowTop - headHeight < viewportTop) {
            viewport.scrollTop = rowTop - headHeight
        } else if (rowBottom > viewportTop + viewportHeight) {
            viewport.scrollTop = rowBottom - viewportHeight
        }
    }

    UNSAFE_componentWillReceiveProps(nextProps) {
        if (nextProps.selected && nextProps.selected !== this.props.selected) {
            this.shouldScrollIntoView = true
        }
    }

    onViewportUpdate() {
        const viewport = ReactDOM.findDOMNode(this)
        const viewportTop = viewport.scrollTop || 0

        const vScroll = calcVScroll({
            viewportTop,
            viewportHeight: viewport.offsetHeight || 0,
            itemCount: this.props.flows.length,
            rowHeight: this.props.rowHeight,
        })

        if (this.state.viewportTop !== viewportTop || !shallowEqual(this.state.vScroll, vScroll)) {
            this.setState({ vScroll, viewportTop })
        }
    }

    render() {
        const { vScroll, viewportTop } = this.state
        const { flows, selected, highlight } = this.props
        const isHighlighted = highlight ? Filt.parse(highlight) : () => false

        return (
            <div className="flow-table" onScroll={this.onViewportUpdate}>
                <table>
                    <thead ref="head" style={{ transform: `translateY(${viewportTop}px)` }}>
                        <FlowTableHead />
                    </thead>
                    <tbody>
                        <tr style={{ height: vScroll.paddingTop }}/>
                        {flows.slice(vScroll.start, vScroll.end).map(flow => (
                            <FlowRow
                                key={flow.id}
                                flow={flow}
                                selected={flow === selected}
                                highlighted={isHighlighted(flow)}
                                onSelect={this.props.selectFlow}
                            />
                        ))}
                        <tr style={{ height: vScroll.paddingBottom }}/>
                    </tbody>
                </table>
            </div>
        )
    }
}

export const PureFlowTable = AutoScroll(FlowTable)

export default connect(
    state => ({
        flows: state.flows.view,
        highlight: state.flows.highlight,
        selected: state.flows.byId[state.flows.selected[0]],
    }),
    {
        selectFlow: flowsActions.select,
    }
)(PureFlowTable)
