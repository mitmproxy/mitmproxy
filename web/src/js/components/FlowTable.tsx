import * as React from "react";
import { connect, shallowEqual } from "react-redux";
import * as autoscroll from "./helpers/AutoScroll";
import { calcVScroll, VScroll } from "./helpers/VirtualScroll";
import FlowTableHead from "./FlowTable/FlowTableHead";
import FlowRow from "./FlowTable/FlowRow";
import Filt from "../filt/filt";
import { Flow } from "../flow";
import { RootState } from "../ducks";

type FlowTableProps = {
    flows: Flow[];
    rowHeight: number;
    highlight: string;
    selected: Flow;
};

type FlowTableState = {
    vScroll: VScroll;
    viewportTop: number;
};

export class PureFlowTable extends React.Component<
    FlowTableProps,
    FlowTableState
> {
    static defaultProps = {
        rowHeight: 32,
    };
    private viewport = React.createRef<HTMLDivElement>();
    private head = React.createRef<HTMLTableSectionElement>();

    constructor(props, context) {
        super(props, context);

        this.state = {
            vScroll: calcVScroll(),
            viewportTop: 0,
        };
        this.onViewportUpdate = this.onViewportUpdate.bind(this);
    }

    componentDidMount() {
        window.addEventListener("resize", this.onViewportUpdate);
        this.onViewportUpdate();
    }

    componentWillUnmount() {
        window.removeEventListener("resize", this.onViewportUpdate);
    }

    getSnapshotBeforeUpdate() {
        return autoscroll.isAtBottom(this.viewport);
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (snapshot) {
            autoscroll.adjustScrollTop(this.viewport);
        }
        this.onViewportUpdate();

        const selectedNewFlow =
            this.props.selected && this.props.selected !== prevProps.selected;
        if (selectedNewFlow) {
            const { rowHeight, flows, selected } = this.props;
            const viewport = this.viewport.current!;
            const head = this.head.current;

            const headHeight = head ? head.offsetHeight : 0;

            const rowTop = flows.indexOf(selected) * rowHeight + headHeight;
            const rowBottom = rowTop + rowHeight;

            const viewportTop = viewport.scrollTop;
            const viewportHeight = viewport.offsetHeight;

            // Account for pinned thead
            if (rowTop - headHeight < viewportTop) {
                viewport.scrollTop = rowTop - headHeight;
            } else if (rowBottom > viewportTop + viewportHeight) {
                viewport.scrollTop = rowBottom - viewportHeight;
            }
            this.onViewportUpdate();
        }
    }

    onViewportUpdate() {
        const viewport = this.viewport.current!;
        const viewportTop = viewport.scrollTop || 0;

        const vScroll = calcVScroll({
            viewportTop,
            viewportHeight: viewport.offsetHeight || 0,
            itemCount: this.props.flows.length,
            rowHeight: this.props.rowHeight,
        });

        if (
            this.state.viewportTop !== viewportTop ||
            !shallowEqual(this.state.vScroll, vScroll)
        ) {
            // the next rendered state may only have much lower number of rows compared to what the current
            // viewportHeight anticipates. To make sure that we update (almost) at once, we already constrain
            // the maximum viewportTop value. See https://github.com/mitmproxy/mitmproxy/pull/5658 for details.
            const newViewportTop = Math.min(
                viewportTop,
                vScroll.end * this.props.rowHeight,
            );
            this.setState({
                vScroll,
                viewportTop: newViewportTop,
            });
        }
    }

    render() {
        const { vScroll, viewportTop } = this.state;
        const { flows, selected, highlight } = this.props;
        const isHighlighted = highlight ? Filt.parse(highlight) : () => false;

        return (
            <div
                className="flow-table"
                onScroll={this.onViewportUpdate}
                ref={this.viewport}
            >
                <table>
                    <thead
                        ref={this.head}
                        style={{ transform: `translateY(${viewportTop}px)` }}
                    >
                        <FlowTableHead />
                    </thead>
                    <tbody>
                        <tr style={{ height: vScroll.paddingTop }} />
                        {flows.slice(vScroll.start, vScroll.end).map((flow) => (
                            <FlowRow
                                key={flow.id}
                                flow={flow}
                                selected={flow === selected}
                                highlighted={isHighlighted(flow)}
                            />
                        ))}
                        <tr style={{ height: vScroll.paddingBottom }} />
                    </tbody>
                </table>
            </div>
        );
    }
}

export default connect((state: RootState) => ({
    flows: state.flows.view,
    highlight: state.flows.highlight,
    selected: state.flows.byId[state.flows.selected[0]],
}))(PureFlowTable);
