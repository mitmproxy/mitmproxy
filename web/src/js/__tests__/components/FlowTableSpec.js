import React from 'react'
import renderer from 'react-test-renderer'
import {PureFlowTable as FlowTable} from '../../components/FlowTable'
import TestUtils from 'react-dom/test-utils'
import { TFlow, TStore } from '../ducks/tutils'
import { Provider } from 'react-redux'

window.addEventListener = jest.fn()

describe('FlowTable Component', () => {
    let selectFn = jest.fn(),
        tflow = TFlow(),
        store = TStore()

    it('should render correctly', () => {
        let provider = renderer.create(
                <Provider store={store}>
                    <FlowTable selectFlow={selectFn} flows={[tflow]}/>
                </Provider>),
            tree = provider.toJSON()
        expect(tree).toMatchSnapshot()
    })

    let provider = renderer.create(
        <Provider store={store} >
            <FlowTable selectFlow={selectFn} flows={[tflow]}/>
        </Provider>),
        flowTable = provider.root.findByType(FlowTable)

    it('should handle componentWillUnmount', () => {
        flowTable.instance.UNSAFE_componentWillUnmount()
        expect(window.addEventListener).toBeCalledWith('resize', flowTable.instance.onViewportUpdate)
    })

    it('should handle componentDidUpdate', () => {
        // flowTable.shouldScrollIntoView == false
        expect(flowTable.instance.componentDidUpdate()).toEqual(undefined)
        // rowTop - headHeight < viewportTop
        flowTable.instance.shouldScrollIntoView = true
        flowTable.instance.componentDidUpdate()
        // rowBottom > viewportTop + viewportHeight
        flowTable.instance.shouldScrollIntoView = true
        flowTable.instance.componentDidUpdate()
    })

    it('should handle componentWillReceiveProps', () => {
        flowTable.instance.UNSAFE_componentWillReceiveProps({selected: tflow})
        expect(flowTable.instance.shouldScrollIntoView).toBeTruthy()
    })
})
