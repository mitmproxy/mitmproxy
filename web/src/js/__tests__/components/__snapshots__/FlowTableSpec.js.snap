// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowTable Component should render correctly 1`] = `
<div
  className="flow-table"
  onScroll={[Function]}
>
  <table>
    <thead
      style={
        Object {
          "transform": "translateY(undefinedpx)",
        }
      }
    >
      <tr>
        <th
          className="col-tls"
          onClick={[Function]}
        >
          
        </th>
        <th
          className="col-icon"
          onClick={[Function]}
        >
          
        </th>
        <th
          className="col-path sort-desc"
          onClick={[Function]}
        >
          Path
        </th>
        <th
          className="col-method"
          onClick={[Function]}
        >
          Method
        </th>
        <th
          className="col-status"
          onClick={[Function]}
        >
          Status
        </th>
        <th
          className="col-size"
          onClick={[Function]}
        >
          Size
        </th>
        <th
          className="col-time"
          onClick={[Function]}
        >
          Time
        </th>
        <th
          className="col-quickactions"
          onClick={[Function]}
        >
          
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        style={
          Object {
            "height": 0,
          }
        }
      />
      <tr
        style={
          Object {
            "height": 0,
          }
        }
      />
    </tbody>
  </table>
</div>
`;
